### PR TITLE
[BE] Pagination for all lists

### PR DESCRIPTION
**Pull request by MihaiDruta**
[BE] Pagination for all lists
https://trello.com/c/0bMghHyt/61-be-pagination-for-all-lists

Implemented pagination for all lists in project.

### CODE REVIEW

File: src/Handlers/ActivityHandler.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        /** @var SerializationContext $context */
+        $context = SerializationContext::create()->setGroups(array('ActivityList'));
+
+        $json = $this->serializer->serialize(
+            $paginatedResults,
+            'json',
+            $context
+        );
+
+        return new JsonResponse([
```
Comment (SpartakusMd) [inline src/Handlers/ActivityHandler.php]: Return only the array. JsonResponse should be returned from the Controller.

File: src/Controller/ActivityController.php [hunk 2]
```diff
@@ -92,26 +115,16 @@ public function __construct(
...
-        );
 
-        return new JsonResponse($json, 200, [], true);
+        return $this->activityHandler->getActivitiesListPaginated($user, $page);
     }
 
     /**
      * Get details about and Activity
-     * @Rest\Get("/{id}")
+     * @Rest\Get("/{id}", requirements={"id"="\d+"})
```
Comment (SpartakusMd) [inline src/Controller/ActivityController.php:135]: Either `{page<\d+>}` or `requirements={"id"="\d+"}` should be used everywhere. Please change it everywhere you will find this.

File: src/Handlers/ActivityHandler.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        $paginator = new Paginator($paginatedResults);
+        $numResults = $paginator->count();
+
+        $paginatedResults = $paginatedResults->getResult();
+
+        /** @var SerializationContext $context */
+        $context = SerializationContext::create()->setGroups(array('ActivityList'));
+
+        $json = $this->serializer->serialize(
+            $paginatedResults,
```
Comment (SpartakusMd) [inline src/Handlers/ActivityHandler.php]: Use `$paginatedResults->getResult()`and remove `$paginatedResults = $paginatedResults->getResult();` from line 39.

File: src/Handlers/ActivityUserHandler.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    private $activityUserRepository;
+
+    public function __construct(SerializerInterface $serializer, ActivityUserRepository $activityUserRepository)
+    {
+
+        $this->serializer = $serializer;
+        $this->activityUserRepository = $activityUserRepository;
+    }
+
+    public function getApplicantsPaginated(Activity $activity, int $page): JsonResponse
```
Comment (SpartakusMd) [inline src/Handlers/ActivityUserHandler.php]: Same comments from ActivityHandler.

File: src/Handlers/ActivityHandler.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            'json',
+            $context
+        );
+
+        return new JsonResponse([
+            'results' => json_decode($json, true),
+            'numResults' => $numResults,
+            'perPage' => Activity::NUM_ITEMS_PER_PAGE,
```
Comment (SpartakusMd) [inline src/Handlers/ActivityHandler.php]: Move the constant to this class.

File: src/Repository/UserRepository.php [hunk 6]
```diff
@@ -47,4 +50,18 @@ public function delete(User $user): void
         $em->remove($user);
         $em->flush();
     }
+
+    public function getApplicantsForActivity(Activity $activity): QueryBuilder
```
Comment (SpartakusMd) [inline src/Repository/UserRepository.php:54]: There is already a method for this: `ActivityUserRepository::getApplicantsForActivity()`

**Review by SpartakusMd — CHANGES_REQUESTED** [review]

File: src/Repository/UserRepository.php (hunk 6 above)
Comment (MihaiDruta) [reply src/Repository/UserRepository.php:54]: The method from ActivityUserRepository didn't work with paginator. So I deleted it and created a new one in UserRepository.

Comment (MihaiDruta) [reply src/Repository/UserRepository.php:54]: https://github.com/doctrine/orm/issues/2596

Comment (SpartakusMd) [reply src/Repository/UserRepository.php:54]: ok

File: src/Controller/ActivityController.php [hunk 7]
```diff
@@ -670,7 +675,7 @@ public function inviteUserToActivity(
 
     /**
      * Get a list of all applicants.
-     * @Rest\Get("/{activityId}/applicants/{page<\d+>}", defaults={"page" = 1})
+     * @Rest\Get("/{activityId}/applicants/{page<\d+>}", defaults={"page" = 1}, requirements={"activityId"="\d+"})
```
Comment (SpartakusMd) [inline src/Controller/ActivityController.php]: Remove `<\d+>`

File: src/Handlers/ActivityHandler.php [hunk 8]
```diff
@@ -2,17 +2,16 @@
...
 use App\Repository\ActivityRepository;
 use Doctrine\ORM\Tools\Pagination\Paginator;
 use JMS\Serializer\SerializationContext;
 use JMS\Serializer\SerializerInterface;
-use Symfony\Component\HttpFoundation\JsonResponse;
-use Symfony\Component\HttpFoundation\Response;
 
 class ActivityHandler
 {
+    public const NUM_ITEMS_PER_PAGE = 4;
```
Comment (SpartakusMd) [inline src/Handlers/ActivityHandler.php]: Use 10 as a default.

File: src/Handlers/ActivityUserHandler.php [hunk 9]
```diff
@@ -3,16 +3,15 @@
...
 use App\Repository\ActivityUserRepository;
 use Doctrine\ORM\Tools\Pagination\Paginator;
 use JMS\Serializer\SerializationContext;
 use JMS\Serializer\SerializerInterface;
-use Symfony\Component\HttpFoundation\JsonResponse;
-use Symfony\Component\HttpFoundation\Response;
 
 class ActivityUserHandler
 {
+    public const NUM_APPLICANTS_PER_PAGE = 5;
```
Comment (SpartakusMd) [inline src/Handlers/ActivityUserHandler.php]: Use 10 as default.

**Review by SpartakusMd — CHANGES_REQUESTED** [review]

**Review by SpartakusMd — APPROVED** [review]

**Review by nspc97 — APPROVED** [review]

**Pull request merged**